### PR TITLE
Add transport-cc extension

### PR DESCRIPTION
**Pull request by adwpc**
Add transport-cc extension

#### Reference issue
https://tools.ietf.org/html/draft-holmer-rmcat-transport-wide-cc-extensions-01#page-3

### CODE REVIEW

File: transportccextension.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+// 0                   1                   2                   3
+// 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+// |       0xBE    |    0xDE       |           length=1            |
+// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+// |  ID   | L=1   |transport-wide sequence number | zero padding  |
+// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type TransportCCExtension struct {
+	ID  uint8
+	L   uint8
```
Comment (maxhawkins) [inline transportccextension.go]: Since L will always equal 1 do you need to set it here?

You could just assert that it's 1 on Unmarshal instead. That will make constructing a TransportCCExtenstion struct easier since you won't need to remember to set L: 1

File: transportccextension.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	TSN uint16
```
Comment (maxhawkins) [inline transportccextension.go]: In general ive tried to avoid using acronyms in structs that require a reader to dig back into an RFC to understand what a packet does.

Could this be TransportSequence or something?

File: transportccextension_test.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+		TSN: 2,
+	}
+
+	if t1 != t2 {
+		t.Error("Unmarshal failed")
+	}
+
+	dstData, _ := t2.Marshal()
+	if !bytes.Equal(dstData, rawData) {
+		t.Error("Marshal failed")
```
Comment (maxhawkins) [inline transportccextension_test.go]: Could you add a case testing an empty packet returns errTooSmall?

Maybe also worth testing a packet that has extra bytes at the end parses correctly.

File: transportccextension.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	// TransportCCExtensionSize 4 byte
+	TransportCCExtensionSize = 4
+)
+
+var (
+	errTooSmall = errors.New("buffer too small")
+)
+
+// TransportCCExtension is a extension payload format in
+// https://tools.ietf.org/html/draft-holmer-rmcat-transport-wide-cc-extensions-01
```
Comment (maxhawkins) [inline transportccextension.go]: Will this format weird in the godoc?

The byte layout is important to have but I'm not sure it belongs in tbe docstring

**Comment by maxhawkins** [comment]
Left you some feedback!

I haven't contributed to the project in a while so I don't want to approve until Sean or someone takes a look, but overall it looks good.

File: transportccextension.go (hunk 1 above)
Comment (at-wat) [reply transportccextension.go]: `L` is a `(length-1)` of the element, in this case two bytes.
`L=3` version of Transport CC extension with extra fields looks been proposed.
https://webrtc.org/experiments/rtp-hdrext/transport-wide-cc-02/
```
  0                   1                   2                   3
  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 |  ID   | L=3   |transport-wide sequence number |T|  seq count  |
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 |seq count cont.|
 +-+-+-+-+-+-+-+-+
```

In my understanding, rfc5285 defines just a way to transmit some pairs of ID and value which are "defined by profile".
So, I'd like to suggest to make it something general like `HeaderExtension` and marshal/unmarshal between binary and `map[ID][]byte`.
How do you think?

Comment (maxhawkins) [reply transportccextension.go]: I've done some thinking along those lines here: https://github.com/pion/rtp/issues/17

Sadly I don't have time to implement it.

I figured we could add that later in a backwards compatible way but do this for now since adam needs this extension immediately.

Comment (adwpc) [reply transportccextension.go]: > Looks nice to have header extensions parser!
> I roughly read rfc5285 and added a comment [#30 (comment)](https://github.com/pion/rtp/pull/30#discussion_r363129755).
> Please let me know if my understanding is wrong :)

New RFC:
https://tools.ietf.org/html/rfc8285

**Comment by adwpc** [comment]
> What's this for? LGTM otherwise

packetizer_test.go import

File: transportccextension.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
+package rtp
+
+import (
+	"encoding/binary"
+	"errors"
+)
+
+const (
+	// TransportCCExtensionSize 4 byte
+	TransportCCExtensionSize = 4
```
Comment (Sean-Der) [inline transportccextension.go]: Does this need to be exported?

**Comment by Sean-Der** [comment]
@adwpc LGTM! Thanks for grabbing, would you mind adding the tests max mentioned and squashing into one commit. Then happy to approve :) sorry I am slow on this.

File: transportccextension.go (hunk 5 above)
Comment (adwpc) [reply transportccextension.go]: Fixed

**Comment by adwpc** [comment]
> @adwpc LGTM! Thanks for grabbing, would you mind adding the tests max mentioned and squashing into one commit. Then happy to approve :) sorry I am slow on this. I had to read the RFC myself, never implemented any of the congestion-control algos myself.

Added test and one commit
